### PR TITLE
initial commit for new blog post

### PR DESCRIPTION
**Pull request by jmimico**
someone please peer review

### CODE REVIEW

**Comment by jmimico** [comment]
Netlify preview site here: https://5c1166444362f20007d493e3--oicr-softeng.netlify.com/

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    info: How we migrated an 8 PB cluster from Filestore to Bluestore
+    image: jared_baker/bluestore/Ceph_Logo_Standard_RGB_120411_fa.png
+header:
+    version: small
+    title: Software Engineering Blog
+    image: jared_baker/le-logo-standard.png
+    icon: icon-blog
+---
+## Introduction
+The Cancer Genome Collaboratory is a Cloud computing environment designed for cancer research and an official repository of 950 TB (and growing) worth of whole genomes for the [ICGC project](https://dcc.icgc.org/repositories). We have used OpenStack and Ceph to provide an extremely scalable and reliable infrastructure. Maintaining the pieces namely through upgrades is our most important and time consuming task in operating a cloud like this. Whether you are chasing new features, bug fixes, security patches or supportability it seems like we are always planning for the next upgrade.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Maybe hyperlink to Collaboratory?  You link to ICGC Repositories later in the paragraph.  Might be useful for novice readers who want to explore Collab.

Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Consider re-wording as _"...official repository of **950 TB worth (and growing)**..."_

Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Hyphenate _"**time-consuming**"_

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+## About our Ceph environment
+Our Ceph cluster was born in 2014 and was built upon Ceph Giant and eventually upgraded to Hammer, then to Ceph Jewel and more recently this year to Luminous. Up until Luminous our OSD's were using the only available osd backend at the time, Filestore, with xfs and co-located journals and 3 replicas for data redundancy. Our number of OSD's has been growing steadily throughout the years to maximize purchase value per terabyte and spreading our budget over a number of years so that hardware purchases only occurred when capacity was forecasted.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Consider writing out full acronym for OSD (Object Storage Daemon) in first occurrence, depending on audience level you're targeting.  For example, I as a newbie, did not know what OSD was reading this post the first time.  However, the post seems to be targeting those highly knowledgeable in Ceph and OpenStack already, so you maybe this comment is irrelevant.
- Consider capitalizing one all lowercase occurrence of 'OSD' to stay consistent throughout post.
- Should 'xfs' be all-caps?

Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Consider capitalizing one all lowercase occurrence of OSD to stay consistent throughout post.

Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Should xfs be capitalized?

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+We currently have 37 storage nodes, each with 36 drives for a total of 1332 OSD's. As mentioned before, purchasing nodes over time means we have a variety of HDD sizes ranging from 4 TB to 12 TB. Our smallest nodes are 144 TB while our largest are 432 TB. Each storage node has 2 separate SSD's for the OS, 256 GB of RAM, dual Xeon CPU's and 40 Gbps of network capability through the use of vlan bonds to partition the traffic type (ceph public, replication, etc).
+
+The primary use case for our Ceph Cluster is the storage of those very large genomic files and are served to researchers using the rados gateway object storage api. We do also supplement the OpenStack environment with volume support using cinder but it amounts to less than 1% of our total utilization of the ceph cluster.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Should 'rados', 'api' be all-caps?
- Should ' cinder' be capitalized?
- Capitalize last occurrence of 'ceph' to be consistent with rest of post

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+## Bluestore in Ceph Luminous
+[Bluestore](https://ceph.com/community/new-luminous-bluestore/) is the new default storage backend starting in Ceph Luminous and proposes to improve write performance by a factor of 2, and sometimes higher in certain scenarios by removing the double write penalty that took place on Filestores journal and xfs partition. Most notably Bluestore removes the requirement of a posix filesystem to store ceph data. See the diagram below for the architectural differences.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Add apostrophe to '_Filestore's_' (should be possessive)
- Should 'xfs' and 'posix' be all-caps?
- Capitalize last occurrence of 'ceph' to be consistent with rest of post

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+<figure>
+    <img id="filestore-vs-bluestore-2" src="{{site.urlimg}}jared_baker/bluestore/filestore-vs-bluestore-2.png" data-featherlight="#filestore-vs-bluestore-2" />
+</figure>
+
+
+## Prequisites
+Since Bluestore was first officially supported in Luminous we needed to upgrade from Ceph Jewel to Luminous. We use Ubuntu 16.04 and the official ceph packages for that distro so it was a straight forward upgrade of adding the repo and upgrading the packages on our mon's, radosgw's and OSD nodes.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: * Capitalize last occurrence of 'ceph' to be consistent with rest of post

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+## Migration Approach
+Drain - Convert - Fill!
+
+There is no 'conversion' from Filestore to Bluestore without major data movement so our approach was to drain a storage node of its data, followed by destroying the OSD and then prepare the OSD's using Bluestore followed by the refilling of the data from the cluster. Rinse and repeat for each storage node in the cluster. Depending on your ceph architecture you may be able to cycle several storage nodes in parallel to reduce the overall migration time. We would migrate 2-3 storage nodes at a time, each in different racks (our failure domain) and tried to keep below a high water mark of 20% 'misplaced objects' during the drain and fills.
+
+In order for this to be successful you have to ensure that your cluster has enough free space to drain each node to account for the temporary loss of cluster capacity when a node has to drain all it's data to the rest of the cluster.
+
+### Draining
+To drain a node we would modify the osd crush weight of all of a storage nodes OSD's to 0. This will trigger an evacuation of data away from this storage node and distribute it to the rest of the cluster.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Add apostrophe to "_storage node's_" (should be possessive)

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+Example, to drain OSD's 720 to 755 we would run:
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Consider re-wording to "_For example, ..._"

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+~~~bash
+for i in $(seq 720 755); do ceph osd crush reweight osd.$i 0; done
+~~~
+
+When the draining is complete you can check ceph's health (ceph -s) to confirm if it is done draining the data. You can also use 'ceph osd df tree' to see how much data the disks still have to transfer off. Our storage nodes took on average 24 hours per server to drain.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Out of curiosity, is 24 hours drain time considered fast, average, or slow, for the amount of data we house?  Any comparison to help the reader understand whether this metric is comparatively fast or slow?

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+ceph-disk zap <osd-id>
+
+# Mark the OSD as destroyed
+ceph osd destroy <osd-id> --yes-i-really-mean-it
+
+# Prepare the disk as Bluestore
+ceph-disk prepare --bluestore /dev/<block-device> --osd-id <osd-id>
+~~~
+
+Notice that when comparing the output of ceph osd tree before/after conversion that the class column may now indicate your osd type (ssd or hdd) which for us, was blank before.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Why was it blank before the conversion for our implementation?  Is this a useful detail to add in for the reader?

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+~~~bash
+for i in $(seq 720 755); do ceph osd crush reweight osd.$i 10.79999; done
+~~~
+
+Depending on the size of the disks you will need to find the right value for your osd crush weight. This is typically the size of the disk in TB. Read more about the differences of ceph osd reweight and ceph osd crush reweight [HERE](https://ceph.com/geen-categorie/difference-between-ceph-osd-reweight-and-ceph-osd-crush-reweight/)
+
+Wait for ceph health to return to HEALTH_OK.
+
+## Tracking and Monitoring
+We use a variety of tools to monitor and report on our OpenStack and Ceph cluster. For this migration we were able to monitor filestores xfs data partition and see the data be copied off to the rest of the cluster.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: Add apostrophe to "_Filestore's_" (should be possessive)

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+<figure>
+    <img id="drain1" src="{{site.urlimg}}jared_baker/bluestore/drain1.png" data-featherlight="#drain1" />
+</figure>
+
+Once the osd is converted to bluestore we no longer have that trusty posix filesystem to monitor in Zabbix so we had to rely on output from ceph mgr to obtain how much data is living on the osd. Grafana and graphite has nice integration with ceph-mgr so that we are still able to see the granularity of data distribution on each storage node
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Add period to end of last sentence

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+<figure>
+    <img id="zabbixvsgrafana" src="{{site.urlimg}}jared_baker/bluestore/zabbixvsgrafana.png" data-featherlight="#zabbixvsgrafana" />
+</figure>
+
+With 37 storage nodes to migrate and each one needing to drain, convert, fill, we had to track it somehow. We are a team of two so we set up a shared google spreadsheet and kept track of the progress.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md:131]: - Is the sheet private/internal OICR only?  Or available to public?  If public, would it be useful to link to the template so others can use?  (I know it's a simple spreadsheet anyone can do, but people tend to be lazy 😄)

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+## Performance impact
+The draining and filling of each storage node causes massive amounts of replication traffic on the cluster but the front end of the ceph cluster was still able to serve the environment unimpacted. Researchers were still able to fetch data from the object storage and openstack volumes continued to work without issue. We also benchmark the download of 100GB genomic files from our object storage every hour and saw no deviation from the normal speed during this migration project.
+<figure>
+    <img id="hourlydownloads" src="{{site.urlimg}}jared_baker/bluestore/hourlydownloads.png" data-featherlight="#hourlydownloads" />
+</figure>
+
+## Issues
+Late in to the project we started to get ceph health warnings, complaining that the monmap was too large. Our cluster is a few years old so our monmap is leveldb. The monmap was growing to to 20GB during the migrations (Under 1GB normally). Leveldb can perform poorly during large data movement events and cause everything to slow down. To remedy this we had to enable mon compaction in our /etc/ceph/ceph.conf and then restart ceph-mon and then wait for the compaction to happen.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Reword "_in to_" as "_into_" (single word)
- Remove duplicate "_to_" in "_The monmap was growing **to to** 20 GB..._"
- Make "_under_" lowercase in "_(Under 1 GB normally)_"

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+~~~bash
+[mon]
+  mon compact on start = true
+~~~
+This wasn't much of an issue since we had allocated 100GB to the partition where the monmap was stored but did result to us looking in to migrating to the monmap backend to Luminous' default of 'rocksdb' which apparently handles data migration events like failures much better.
```
Comment (b-f-chan) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Add apostrophe to "_wasn't_" (should be possessive)
- Consider rewording to "_...did result **in** us looking **into** migrating..._" (2 changes proposed here)

**Review by b-f-chan — CHANGES_REQUESTED** [review]
Hi @jmimico, I did an editorial review and posted comments directly into the 2018-12-15-Bluestore.md file.

However, I noticed you had a later commit with some edits from George (apologies for not seeing the latest commit).  But only one of your changes / his comments overlaps with mine.

Sorry for the picky comments (I used to be a technical writer 10 years ago 😛), they're just suggestions, so feel free to disregard any you feel are irrelevant or onerous.  Thanks.

File: _posts/jared_baker/2018-12-15-Bluestore.md (hunk 3 above)
Comment (Fgerthoffert) [reply _posts/jared_baker/2018-12-15-Bluestore.md]: > is the storage of those very large genomic files and are served to researchers
=> 
> is the storage of those very large genomic files which are served to researchers

File: _posts/jared_baker/2018-12-15-Bluestore.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+### Filling
+Now it's time to populate the bluestore OSD's with data. Just like the draining step we will change the osd crush weight
+
+Example, to fill OSD's 720 to 755 (12TB disks) we would run:
+~~~bash
+for i in $(seq 720 755); do ceph osd crush reweight osd.$i 10.79999; done
+~~~
+
+Depending on the size of the disks you will need to find the right value for your osd crush weight. This is typically the size of the disk in TB. Read more about the differences of ceph osd reweight and ceph osd crush reweight [HERE](https://ceph.com/geen-categorie/difference-between-ceph-osd-reweight-and-ceph-osd-crush-reweight/)
```
Comment (Fgerthoffert) [inline _posts/jared_baker/2018-12-15-Bluestore.md]: - Dot at the end of the sentence
- I would not capitalize HERE, but that's personal preferences

**Review by Fgerthoffert — COMMENTED** [review]
Well written blog post

**Comment by jmimico** [comment]
I've addressed the items. Thanks for the feedback everyone!

**Review by jmimico — COMMENTED** [review]
Updated in commits

**Review by b-f-chan — APPROVED** [review]

**Pull request merged**